pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,43 +1207,69 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
 		return "", err
 	}
 
-	// vmdks will be created inside kubevols directory
-	kubeVolsPath := filepath.Clean(ds.Path(VolDir)) + "/"
-	err = makeDirectoryInDatastore(vs.client, dc, kubeVolsPath, false)
-	if err != nil && err != ErrFileAlreadyExist {
-		glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
-		return "", err
-	}
-	glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
+	// Create a disk with the VSAN storage capabilities specified in the volumeOptions.StorageProfileData.
+	// This is achieved by following steps:
+	// 1. Create dummy VM if not already present.
+	// 2. Reconfigure the VM with the disk.
+        // 3. Detach the disk from the dummy VM.
+	if volumeOptions.StorageProfileData != "" {
+		// Check if the datastore is VSAN if any capability requirements are specified.
+		// VSphere cloud provider now only supports VSAN capabilities requirements
+		ok, err := checkIfDatastoreTypeIsVSAN(vs.client, ds)
+		if err != nil {
+			return "", fmt.Errorf("Failed while determining whether the datastore: %q"+
+				" is VSAN or not.", datastore)
+		}
 
-	vmDiskPath := kubeVolsPath + volumeOptions.Name + ".vmdk"
+		if !ok {
+			return "", fmt.Errorf("The specified datastore: %q is not a VSAN datastore."+
+				" The policy parameters will work only with VSAN Datastore."+
+				" So, please specify a valid VSAN datastore", datastore)
+		}
 
-	// Create a virtual disk manager
-	virtualDiskManager := object.NewVirtualDiskManager(vs.client.Client)
+		// Check if the VM exists.
+		vmRegex := vs.cfg.Global.WorkingDir + DummyVMName
+		dummyVM, err := f.VirtualMachine(ctx, vmRegex)
+		if err != nil {
```

Comment:
The operation made on the VC to this VM would be serialized. Even though 2 requests attempting to create a dummyVM comes, the second one would fail and would eventually succeed next time (uses pre created VM) as Kubernetes makes multiple requests to createVolume until PVC is bound.